Add basic support for arrow functions

NickTomlin opened the pull request:
This builds on some of the work in https://github.com/TroyAlford/react-jsx-parser/pull/114 to add support for _very_ simple arrow functions. I've wanted this to allow users to do simple templating on collections without having to layer on an additional template language inside JSX.



### Remaining questions

#### Do we need "full" function support?

I chose to parse the body of the function, rather than try to eval it. There are a few reasons for this:

1. It was relatively easy and followed the parse and return grain already laid down
2. Adding `eval`/`Function("alert('hi')")` into the mix is a little scary 🙀 

I think it might actually be fine to _not_ allow full function bodies and just support the simplified expression syntax if we wanted. I don't see this library wanting to necessarily support arbitrary JS parsing and that sort of thing is probably best left to a handler function.

I'm new to `acorn` and this library so please let me know if i've missed something obvious.

NickTomlin committed:
feat: Add basic support for arrow functions in JSX

NickTomlin commented on source/types/acorn-jsx.d.ts:
```diff
@@ -17,6 +17,22 @@ declare module 'acorn-jsx' {
+  export interface ArrowFunctionExpression extends BaseFunctionExpression {
+		type: 'ArrowFunctionExpression';
+	}
+	export interface FunctionExpression extends BaseFunctionExpression {
```
I left this in pending discussion around support for "full" functions (e.g. arrow with braces/return and "normal" function expression). 

I can remove this if we decide simple arrows are the way to go.

TroyAlford replied:
My vote is that simple arrows are the way to go because it keeps things simpler.

NickTomlin commented on source/components/JsxParser.tsx:
```diff
@@ -73,26 +74,26 @@ export default class JsxParser extends React.Component<TProps> {
 	}
 
-	#parseExpression = (expression: AcornJSX.Expression): any => {
+	#parseExpression = (expression: AcornJSX.Expression, scope?: Scope): any => {
```
I don't feel particularly wonderful about adding `scope` everywhere, but i'm not sure of a better way to allow access to the function params in subsequent expressions.

TroyAlford replied:
Agreed - it feels a little odd. At the same time, this entire thing is a recursive algorithm, parsing a tree - so allowing the parent context to pass along the current scope seems reasonable. It's essentially allowing the sap to run from the trunk of the tree down the branches, which is fine.

NickTomlin committed:
feat: Add scope support to JSX attributes within functions

NickTomlin committed:
chore: Restructure tests; add coverage for spread props

TroyAlford commented on source/components/JsxParser.test.tsx:
```diff
@@ -1238,4 +1238,76 @@ describe('JsxParser Component', () => {
+					jsx="{items.map(item => <p><PropTest booleanAttribute={item.friend} /></p>)}"
+				/>,
+			)
+			expect(html).toEqual('<p>val:true</p><p>val:false</p>')
```
Should probably also assert, here, that the `booleanAttribute` prop is actually set to a boolean value. I think technically, even if something in the binding, here, converted the binding to a string, the test would pass as-written. (I don't think it _is_ doing that, just saying it might be a stronger test case)

TroyAlford commented on source/components/JsxParser.test.tsx:
```diff
@@ -1238,4 +1238,76 @@ describe('JsxParser Component', () => {
+
+		it.skip('[NOT IMPLEMENTED: included for PR discussion] supports function expressions with nested jsx', () => {
+			// this doesn't work because we need to support
+			// ReturnStatement + BlockStatement
```
Yeah - I think that starting down this rabbit-hole, as you mentioned, begins the perilous journey into parsing JS instead of JSX, which we should avoid. The inline arrow-function style binding is great, because it allows the `.map()` but this opens a can of worms.

We originally discussed having this style be acceptable, as long as the _only_ statement inside the body of the function was a return statement - at which point it's functionally equivalent to the above test-cases, but IMHO it's simpler to just leave it out and ONLY allow the arrow-function syntax with no body.

TroyAlford commented on source/components/JsxParser.tsx:
```diff
@@ -119,15 +120,21 @@ export default class JsxParser extends React.Component<TProps> {
 		case 'Identifier':
+			if (scope?.[expression.name]) {
+				return scope?.[expression.name]
+			}
```
That's brilliant - I love the way you've handled this case.

NickTomlin replied:
😅 I realized that this would fail for falsy scope values so I had to update it. The pattern would be a nice fit for the statement below though!

TroyAlford commented on source/components/JsxParser.tsx:
```diff
@@ -162,10 +169,22 @@ export default class JsxParser extends React.Component<TProps> {
 			}
 			return undefined
+		case 'ArrowFunctionExpression':
+		case 'FunctionExpression':
```
Do we need to support `FunctionExpression` at all, if we're only allowing the `() => value` non-body syntax?

NickTomlin committed:
fix: Use "in" operator to ensure we do not miss falsy attributes on scope

NickTomlin committed:
fix: Update tests to ensure attributes are passed through with the correct type

NickTomlin committed:
chore: Clean up typings; fix eslint

NickTomlin commented:
Thanks for the feedback! I've updated the PR to focus on the simplified arrow expression. If this sounds good I can also add a note in the docs. Edit: added docs.

NickTomlin committed:
chore: Update docs

TroyAlford commented:
@NickTomlin this is now released with v1.29.0 :) Thank you for contributing such a valuable feature!

NickTomlin commented:
My pleasure, @TroyAlford. Thank you for this library!
